### PR TITLE
Implement simplified backend selection

### PR DESCRIPTION
**Pull request by tarcieri**
As proposed in #414, this commit changes the backend selection approach, introspecting `target_pointer_width` to select `u32_backend` vs `u64_backend` (or `fiat_u32_backend`/`fiat_u64_backend` if the `fiat_backend` feature is enabled).

This helps eliminate the use of non-additive features, and also the rather confusing errors that happen if multiple backends are selected (i.e. thousands of lines of rustc errors).

The selection logic checks if `target_pointer_width = "64"` and uses the 64-bit backend, or falls back to the 32-bit backend otherwise. This means the crate will always have a valid backend regardless of the pointer width, although there may be odd edge cases for exotic platforms which would optimally use the 64-bit backend but have a non-"64" target pointer width for whatever reason. We can handle those cases as they come up.

### CODE REVIEW

**Comment by tarcieri** [comment]
Note: I deliberately punted on trying to change `fiat_backend` and `simd_backend` into a `cfg` attribute to keep this PR relatively simple

File: Cargo.toml [hunk 1]
```diff
@@ -43,6 +43,7 @@ name = "dalek_benchmarks"
 harness = false
 
 [dependencies]
+cfg-if = "1"
```
Comment (tarcieri) [inline Cargo.toml:46]: Note this is already a transitive hard dependency via the `sha2` crate which uses it in a similar capacity.

It's maintained by the core team, so nothing to worry about as far as supply chain attacks go.

File: README.md [hunk 2]
```diff
@@ -99,18 +99,6 @@ Curve arithmetic is implemented using one of the following backends:
...
-```sh
-cargo build --no-default-features --features "std u32_backend"
-cargo build --no-default-features --features "std u64_backend"
-# Requires nightly, RUSTFLAGS="-C target_feature=+avx2" to use avx2
-cargo build --no-default-features --features "std simd_backend"
-# Requires nightly, RUSTFLAGS="-C target_feature=+avx512ifma" to use ifma
-cargo build --no-default-features --features "std simd_backend"
-```
-Crates using `curve25519-dalek` can either select a backend on behalf of their
-users, or expose feature flags that control the `curve25519-dalek` backend.
```
Comment (tarcieri) [inline README.md:112]: Could probably use a description of the new backend selection logic here.

File: README.md [hunk 3]
```diff
@@ -166,8 +154,7 @@ compiled with appropriate `target_feature`s, so this cannot occur.
 Benchmarks are run using [`criterion.rs`][criterion]:
 
 ```sh
-cargo bench --no-default-features --features "std u32_backend"
-cargo bench --no-default-features --features "std u64_backend"
+cargo bench --no-default-features --features std
 # Uses avx2 or ifma only if compiled for an appropriate target.
 export RUSTFLAGS="-C target_cpu=native"
 cargo bench --no-default-features --features "std simd_backend"
```
Comment (rozbb) [inline README.md:160]: `cargo bench --features "simd_backend"` will suffice

File: src/lib.rs [hunk 4]
```diff
@@ -105,18 +105,6 @@
 //! * an `avx2` backend using [parallel formulas][parallel_doc] and `avx2` instructions (sets speed records);
```
Comment (rozbb) [inline src/lib.rs:105]: All this is killed by #429

**Review by rozbb — CHANGES_REQUESTED** [review]
Looking good! And nice changes to the CI. Left some small comments here and there. After that, and a conflict resolution, this is good to go.

File: README.md (hunk 3 above)
Comment (tarcieri) [reply README.md]: FYI, running that I notice that backend no longer compiles, but it's unrelated to this PR.

File: README.md [hunk 5]
```diff
@@ -99,18 +99,6 @@ Curve arithmetic is implemented using one of the following backends:
 * an `avx2` backend using [parallel formulas][parallel_doc] and `avx2` instructions (sets speed records);
 * an `ifma` backend using [parallel formulas][parallel_doc] and `ifma` instructions (sets speed records);
```
[the remark this answers is not in the document]
Comment (tarcieri) [reply README.md:104]: Added in f0bff96

File: src/lib.rs (hunk 4 above)
Comment (tarcieri) [reply src/lib.rs]: Removed in rebase

File: README.md (hunk 2 above)
Comment (tarcieri) [reply README.md:112]: Added in f0bff96

File: README.md [hunk 6]
```diff
@@ -166,8 +154,7 @@ compiled with appropriate `target_feature`s, so this cannot occur.
 Benchmarks are run using [`criterion.rs`][criterion]:
 
 ```sh
-cargo bench --no-default-features --features "std u32_backend"
-cargo bench --no-default-features --features "std u64_backend"
+cargo bench --no-default-features --features std
```
[the remark this answers is not in the document]
Comment (tarcieri) [reply README.md]: Fixed in f0bff96

File: README.md (hunk 3 above)
Comment (tarcieri) [reply README.md]: Fixed in f0bff96

**Pull request merged**